### PR TITLE
port glu to structured kernel

### PR DESCRIPTION
**Pull request by makslevental**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61433**

Differential Revision: [D29626288](https://our.internmc.facebook.com/intern/diff/D29626288)

### CODE REVIEW

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

File: aten/src/ATen/native/GatedLinearUnit.cpp
Comment (makslevental) [inline aten/src/ATen/native/GatedLinearUnit.cpp:40]: @ezyang is there a better way to do this? i.e. not recompute

File: aten/src/ATen/native/GatedLinearUnit.cpp
Comment (makslevental) [inline aten/src/ATen/native/GatedLinearUnit.cpp:23]: I'm not sure if we should have `build_unary_op(maybe_get_output(), self);` here after setting output. I had it here before but since we don't have `structured_inherits: TensorIteratorBase` in `native_functions.yaml` I had to remove it. Note that `TensorIteratorBase` can't be used (I think?) since `DECLARE_DISPATCH(activation_fn, glu_stub);` points to `using activation_fn = void (*)(TensorIterator&);`

Comment (makslevental) [reply aten/src/ATen/native/GatedLinearUnit.cpp:23]: also not sure if this is the right call to `set_output`; i see the 3 arg call in some places but I couldn't get type coercion/overloading to work for some reason when writing 

`set_output(0, newSizes, self.options())`

i.e. `newSizes` wasn't being coerced (cast?) to `IntArrayRef`(again that's my hypothesis for why 3 arg call didn't type check).

**Comment by makslevental** [comment]
I tested this by running [TEST_F(FunctionalTest, GLU)](https://github.com/pytorch/pytorch/blob/master/test/cpp/api/functional.cpp#L981).

File: aten/src/ATen/native/native_functions.yaml
Comment (makslevental) [inline aten/src/ATen/native/native_functions.yaml:8332]: not sure if this is correct. reading the docs on structured native functions (in particular `device_check`) it seems like `TensorIterator::borrowing_binary_op` should be performing this check already.

File: aten/src/ATen/native/native_functions.yaml
Comment (desertfire) [inline aten/src/ATen/native/native_functions.yaml:8325]: Do we need "structured_inherits: TensorIteratorBase" here?

Comment (makslevental) [reply aten/src/ATen/native/native_functions.yaml:8325]: i'm not sure but currently it's not possible due to

[`DECLARE_DISPATCH(activation_fn, glu_stub);`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.h#L54) and [`using activation_fn = void (*)(TensorIterator&);
`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.h#L16) in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.h

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

File: aten/src/ATen/native/GatedLinearUnit.cpp
Comment (desertfire) [inline aten/src/ATen/native/GatedLinearUnit.cpp:22]: It seems like `newSizes` is not used at all?

**Review by desertfire — APPROVED** [review]
LGTM. @ezyang, any comments?

File: aten/src/ATen/native/GatedLinearUnit.cpp
Comment (makslevental) [reply aten/src/ATen/native/GatedLinearUnit.cpp:22]: ah right that was from before (when I was using `set_output`). I'll remove but I would like someone to confirm for me that indeed the substitution `set_output -> build_borrowing_binary_op` is correct.

File: aten/src/ATen/native/GatedLinearUnit.cpp
Comment (desertfire) [inline aten/src/ATen/native/GatedLinearUnit.cpp:8]: Any reason you decided to change `glu_out` to `glu`? I think here the implementation is meant for `glu_out` and `glu` is meant to leverage this method.

**Review by desertfire — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/GatedLinearUnit.cpp
Comment (makslevental) [reply aten/src/ATen/native/GatedLinearUnit.cpp:8]: That is the pattern here: https://github.com/pytorch/pytorch/pull/57619/files and indeed `TORCH_META_FUNC(glu_out)` does not correspond to a header declaration in `NativeMetaFunctions.h`. I am not sure if this is correct or if https://github.com/pytorch/pytorch/pull/57619/files is in error as well. @ezyang can probably advise.

Comment (ezyang) [reply aten/src/ATen/native/GatedLinearUnit.cpp:8]: Defining the meta func for glu and not glu_out is correct!

File: aten/src/ATen/native/GatedLinearUnit.cpp
Comment (ezyang) [reply aten/src/ATen/native/GatedLinearUnit.cpp:22]: Check if tests pass, but I think this is OK because the narrowed firstHalf/secondHalf tensors will induce TensorIterator to produce the correct size.

**Comment by ezyang** [comment]
lgtm

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626288).

**Review by Krovatkin — APPROVED** [review]
Looks gut! Unfortunately, I'm not super familiar with structured kernels, so maybe you might want to wait for one more review.

**Pull request closed without merging**
